src/main/scala/com/evolutiongaming/crypto/Crypto.scala

```diff
@@ -1,149 +1,225 @@
     *
-    * The provider used is by default this uses the platform default JSSE provider.  This can be overridden by defining
-    * `play.crypto.provider` in `application.conf`.
+    * If the current [[encryptAES]] algorithm is used, it is guaranteed that if a decrypted value is returned
+    * it is the original one and the private key is valid. In case a wrong private key is used, an exception
+    * will be thrown.
     *
-    * The transformation used is by default `AES/CTR/NoPadding`.  It can be configured by defining
-    * `play.crypto.aes.transformation` in `application.conf`.  Although any cipher transformation algorithm can
-    * be selected here, the secret key spec used is always AES, so only AES transformation algorithms will work.
-    *
-    * @param value      An hexadecimal encrypted string.
-    * @param privateKey The key used to encrypt.
-    * @return The decrypted String.
+    * @param value      an encrypted string produced by the [[encryptAES]] method
+    * @param privateKey private key string used in encryption
+    * @return decrypted string
     */
   def decryptAES(value: String, privateKey: String): String = {
-    val seperator = "-"
-    val sepIndex = value.indexOf(seperator)
+    val separator = "-"
+    val sepIndex = value.indexOf(separator)
     if (sepIndex < 0) {
-      decryptAESVersion0(value, privateKey)
+      AES_V0.decrypt(value, privateKey)
     } else {
-      val version = value.substring(0, sepIndex)
-      val data = value.substring(sepIndex + 1, value.length())
+      val version = value.take(sepIndex)
+      val data = value.drop(sepIndex + 1)
       version match {
         case "1" =>
-          decryptAESVersion1(data, privateKey)
+          AES_V1.decrypt(data, privateKey)
         case "2" =>
-          decryptAESVersion2(data, privateKey)
-        case _ =>
+          AES_V2.decrypt(data, privateKey)
+        case "3" =>
+          AES_V3.decrypt(data, privateKey)
+        case _   =>
           throw new RuntimeException("Unknown version")
       }
     }
   }
 
-  private def validateAesKeyLength(key: String): Unit =
-    require(key.getBytes("utf-8").length <= AesKeyBytesMaxSize, throw new AesKeyTooLong)
+  /** AES legacy V0 (no versioning) mode support - it has restrictions on key size */
+  private object AES_V0 {
+    private val CipherAlgorithm = "AES"
+    private val CipherTransformation = "AES"
+
+    def decrypt(value: String, privateKey: String): String = {
+      val privateKeyBytes = privateKey.getBytes(UTF_8)
+      require(privateKeyBytes.length <= AesKeyBytesMaxSize, throw new AesKeyTooLong)
+      val effectiveSecretKey = privateKeyBytes.take(AesKeyBytesMaxSize)
+      val skeySpec = new SecretKeySpec(effectiveSecretKey, CipherAlgorithm)
+      val cipher = Cipher.getInstance(CipherTransformation)
+      cipher.init(Cipher.DECRYPT_MODE, skeySpec)
+      new String(cipher.doFinal(Hex.decodeHex(value)))
```

Comment:
could we have method call per line? so in case of exceptions it is visible which failed ?

Reply:
Agreed.

Reply:
Fixed all the similar places I found